lib/spack/spack/build_environment.py

```diff
@@ -160,11 +159,28 @@ def set_compiler_environment_variables(pkg, env):
     env.set('SPACK_F77_RPATH_ARG', compiler.f77_rpath_arg)
     env.set('SPACK_FC_RPATH_ARG',  compiler.fc_rpath_arg)
 
-    # Add every valid compiler flag to the environment, prefixed with "SPACK_"
+    # Trap spack-tracked compiler flags as appropriate.
+    # env_flags are easy to accidentally override.
+    inject_flags = {}
+    env_flags = {}
+    build_system_flags = {}
+    for flag in spack.spec.FlagMap.valid_compiler_flags():
+        injf, envf, bsf = pkg.flag_handler(flag, pkg.spec.compiler_flags[flag])
+        inject_flags[flag] = [] if injf is None else injf
```

Comment:
IMO it would be slightly more idiomatic to say:

```
inject_flags[flag] = inf or []
```

(whenever I see tests for `None` specifically I have to think about why it is being treated specially)

Reply:
I didn't realize I could do that, yes that's definitely cleaner!